examples/autodiff.jl

```diff
@@ -0,0 +1,146 @@
+# # AutoDiff API
+
+# The goal of this tutorial is to give users already familiar with automatic
+# differentiation (AD) an overview
+# of the Enzyme differentiation API for the following differentiation modes
+# * Reverse mode
+# * Forward mode
+# * Forward over reverse mode
+# * Vector Forward over reverse mode
+# # Defining a function
+# Enzyme differentiates arbitrary multivariate vector functions as the most
+# general case in automatic differentiation
+# ```math
+# f: \mathbb{R}^n \rightarrow \mathbb{R}^m, y = f(x)
+# ```
+# For simplicity we define a vector function with ``m=1``. However, this
+# tutorial can easily be applied to arbitrary ``m \in \mathbb{N}``.
+using Enzyme
+
+function f(x::Array{Float64}, y::Array{Float64})
+    y[1] = x[1] * x[1] + x[2] * x[1]
+    return nothing
+end;
+
+# # Reverse mode
+# The reverse model in AD is defined as
+# ```math
+# \begin{aligned}
+# y &= f(x) \\
+# \bar{x} &= \bar{y} \cdot \nabla f(x)
+# \end{aligned}
+# ```
+# bar denotes an adjoint variable. Note that executing an AD in reverse mode
+# computes both ``y`` and the adjoint ``\bar{x}``.
+x  = [2.0, 2.0]
+bx = [0.0, 0.0]
+y  = [0.0]
+by = [1.0];
+
+# Enzyme stores the value and adjoint of a variable in an object of type
+# `Duplicated` where the first element represent the value and the second the
+# adjoint. Evaluating the reverse model using Enzyme is done via the following
+# call.
+Enzyme.autodiff(f, Duplicated(x, bx), Duplicated(y, by));
+# This yields the gradient of `f` in `bx` at point `x = [2.0, 2.0]`. `by` is called the seed and has
+# to be set to ``1.0`` in order to compute the gradient. Let's save the gradient for later.
+g = copy(bx)
+
+# # Forward mode
+# The forward model in AD is defined as
+# ```math
+# \begin{aligned}
+# y &= f(x) \\
+# \dot{y} &= \nabla f(x) \cdot x
+# \end{aligned}
+# ```
+# To obtain the first element of the gradient using the forward model we have to
+# seed ``\dot{x}`` with ``\dot{x} = [1.0,0.0]``
+x  = [2.0, 2.0]
+dx = [1.0, 0.0]
+y  = [0.0]
+dy = [0.0];
+# In the forward mode the second element of `Duplicated` stores the tangent.
+Enzyme.autodiff(Forward, f, Duplicated(x, dx), Duplicated(y, dy));
+
+# We can now verify that indeed the reverse mode and forward mode yield the same
+# result for the first component of the gradient. Note that to acquire the full
+# gradient one needs to execute the forward model a second time with the seed
+# `dx` set to `[0.0,1.0]`.
+
+# Let's verify whether the reverse and forward model agree.
+g[1] ≈ dy[1]
```

Comment:
Can we introduce a short paragraph to explain how to use vectorized forward mode here? This could be useful to explain how to evaluate sparse Jacobians, and it would be a nice increment to explain the vectorized forward over reverse later.

Reply:
👍I will add a sparse Jacobian computation using vector forward and using the Symbolics.jl sparsity detection and SparseDiffTools.jl coloring.

Reply:
I'm working on a sparse example. Let's merge this first and do the batched forward in a separate one. I think it may make this one too heavy.